### PR TITLE
Add script and workflow to generate unsigned release artifacts

### PR DESCRIPTION
**Pull request by madebr**
Example output: https://github.com/madebr/SDL/actions/runs/5756398589
It generates:
- the sources (`SDL3-3.0.0.zip` and `SDL3-3.0.0.tar.gz`)
- msvc binaries (`SDL3-3.0.0-win32-x86.zip`, `SDL3-3.0.0-win32-x64.zip` and `SDL3-3.0.0-devel-3.0.0-VC.zip`)
- apple binaries (`SDL3-3.0.0.dmg`)

The `build-scripts/releaser.py` script generates the archives, and should also run on non-GitHub systems.
The artifacts are built against, and on some platforms the tests are also run.
(SDL3.dmg does not provide SDL3_test :cry: )

The workflow has a `workflow_dispatch` trigger, which means it needs to be started explicitly and passed arguments to.
Using `gh`, it starts with:
```
gh workflow run release.yml --ref main -f commit=fab5e6ece20971ab8f82198b2381031ed9e7c0a5
```
where:
- `main` is the branch that contains the `releaser.yml` workflow that needs to be executed, and that contains the `releaser.py` script
- `fab5e6ece20971ab8f82198b2381031ed9e7c0a5` is the commit hash that needs a release

@slouken 
The xcode project should be updated to put `sdl3-config.cmake` and `sdl3-config-version.cmake` in the dmg image only once, at `/share/cmake/SDL3`.

## Existing Issue(s)
#8065

### CODE REVIEW

File: Xcode/SDLTest/SDLTest.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -1309,6 +1309,7 @@
 		DB445EFA18184BB600B306B0 /* testdropfile.c */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.c; path = testdropfile.c; sourceTree = "<group>"; };
 		DB89957E18A19ABA0092407C /* testhotplug.app */ = {isa = PBXFileReference; explicitFileType = wrapper.application; includeInIndex = 0; path = testhotplug.app; sourceTree = BUILT_PRODUCTS_DIR; };
 		DB89958318A19B130092407C /* testhotplug.c */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.c; path = testhotplug.c; sourceTree = "<group>"; };
+		DBBC552C182831D700F3CA8D /* TestDropFile-Info.plist */ = {isa = PBXFileReference; lastKnownFileType = text.plist.xml; path = "TestDropFile-Info.plist"; sourceTree = SOURCE_ROOT; };
```
Comment (slouken) [inline Xcode/SDLTest/SDLTest.xcodeproj/project.pbxproj:1312]: This is wrong, you should drop the changes in this project as well as the addition of TestDropFile-Info.plist, which was removed in main.

**Comment by slouken** [comment]
> The xcode project should be updated to put sdl3-config.cmake and sdl3-config-version.cmake in the dmg image only once, at /share/cmake/SDL3.

I can do this once this is merged.

File: Xcode/SDLTest/SDLTest.xcodeproj/project.pbxproj (hunk 1 above)
Comment (madebr) [reply Xcode/SDLTest/SDLTest.xcodeproj/project.pbxproj]: Oops. Don't know how I did this.

**Comment by slouken** [comment]
I'm planning on creating a pre-release of SDL 3.0 soon, let's revisit this?

**Comment by madebr** [comment]
Im seeing [this error](https://github.com/madebr/SDL/actions/runs/8398951342/job/23004565705#step:6:33) when building the dmg.
Can a mac person show me what I am doing wrong?
```
  [INFO] Executing args=['xcodebuild', '-project', PosixPath('/private/tmp/tardir/SDL3-3.0.0/Xcode/SDL/SDL.xcodeproj'), '-target', 'SDL3.dmg', '-configuration', 'Release']
  commit          = 2b9159a98a388d4a88a4ada16c991dceb43dfdf6
  out             = /Users/runner/work/SDL/SDL/dist
  actions         = {'xcframework'}
  dry             = False
  force           = False
  cmake_generator = Ninja
Command line invocation:
    /Applications/Xcode_14.2.app/Contents/Developer/usr/bin/xcodebuild -project /private/tmp/tardir/SDL3-3.0.0/Xcode/SDL/SDL.xcodeproj -target SDL3.dmg -configuration Release

User defaults from command line:
    IDEPackageSupportUseBuiltinSCM = YES

2024-03-23 03:07:57.320 xcodebuild[966:4846] CFPropertyListCreateFromXMLData(): Old-style plist parser: missing semicolon in dictionary on line 2636. Parsing will be abandoned. Break on _CFPropertyListMissingSemicolon to debug.
2024-03-23 03:07:57.322 xcodebuild[966:4846] Error Domain=NSCocoaErrorDomain Code=3840 "JSON text did not start with array or object and option to allow fragments not set. around line 1, column 0." UserInfo={NSDebugDescription=JSON text did not start with array or object and option to allow fragments not set. around line 1, column 0., NSJSONSerializationErrorIndex=0}
2024-03-23 03:07:57.326 xcodebuild[966:4846] Writing error result bundle to /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/ResultBundle_2024-23-03_03-07-0057.xcresult
xcodebuild: error: Unable to read project 'SDL.xcodeproj' from folder '/private/tmp/tardir/SDL3-3.0.0/Xcode/SDL'.
	Reason: The project ‘SDL’ is damaged and cannot be opened due to a parse error. Examine the project file for invalid edits or unresolved source control conflicts.

Path: /tmp/tardir/SDL3-3.0.0/Xcode/SDL/SDL.xcodeproj
```

This is the only diff applied to the Xcode project:
```diff
diff --git a/Xcode/SDL/SDL.xcodeproj/project.pbxproj b/Xcode/SDL/SDL.xcodeproj/project.pbxproj
--- a/Xcode/SDL/SDL.xcodeproj/project.pbxproj
+++ b/Xcode/SDL/SDL.xcodeproj/project.pbxproj
@@ -2633,7 +2633,7 @@
                        );
                        runOnlyForDeploymentPostprocessing = 0;
                        shellPath = /bin/sh;
-                       shellScript = "mkdir -p build/dmg-tmp\ncp -a build/SDL3.xcframework build/dmg-tmp/\n\ncp pkg-support/resources/License.txt build/dmg-tmp\ncp pkg-support/resources/ReadMe.txt build/dmg-tmp\n\n# remove the .DS_Store files if any (we may want to provide one in the future for fancy .dmgs)\nfind build/dmg-tmp -name .DS_Store -exec rm -f \"{}\" \\;\n\n# for fancy .dmg\nmkdir -p build/dmg-tmp/.logo\ncp pkg-support/resources/SDL_DS_Store build/dmg-tmp/.DS_Store\ncp pkg-support/sdl_logo.pdf build/dmg-tmp/.logo\n\n# create the dmg\nhdiutil create -ov -fs HFS+ -volname SDL3 -srcfolder build/dmg-tmp build/SDL3.dmg\n\n# clean up\nrm -rf build/dmg-tmp\n";
+                       shellScript = "mkdir -p build/dmg-tmp/share/cmake/SDL3\ncp -a build/SDL3.xcframework build/dmg-tmp/\n\ncp pkg-support/resources/License.txt build/dmg-tmp\ncp pkg-support/resources/ReadMe.txt build/dmg-tmp\ncp pkg-support/resources/CMake/sdl3-config.cmake build/dmg-tmp/share/cmake/SDL3\ncp pkg-support/resources/SDL3/sdl3-config-version.cmake build/dmg-tmp/share/cmake/SDL3\n\n\nmkdir -p \n\n# remove the .DS_Store files if any (we may want to provide one in the future for fancy .dmgs)\nfind build/dmg-tmp -name .DS_Store -exec rm -f "{}" \\;\n\n# for fancy .dmg\nmkdir -p build/dmg-tmp/.logo\ncp pkg-support/resources/SDL_DS_Store build/dmg-tmp/.DS_Store\ncp pkg-support/sdl_logo.pdf build/dmg-tmp/.logo\n\n# create the dmg\nhdiutil create -ov -fs HFS+ -volname SDL3 -srcfolder build/dmg-tmp build/SDL3.dmg\n\n# clean up\nrm -rf build/dmg-tmp\n";
                };
                F3B38CF0296F63D1005DA6D3 /* ShellScript */ = {
                        isa = PBXShellScriptBuildPhase;
```

**Comment by slouken** [comment]
> Im seeing [this error](https://github.com/madebr/SDL/actions/runs/8398951342/job/23004565705#step:6:33) when building the dmg. Can a mac person show me what I am doing wrong?
> 
> ```
>   [INFO] Executing args=['xcodebuild', '-project', PosixPath('/private/tmp/tardir/SDL3-3.0.0/Xcode/SDL/SDL.xcodeproj'), '-target', 'SDL3.dmg', '-configuration', 'Release']
>   commit          = 2b9159a98a388d4a88a4ada16c991dceb43dfdf6
>   out             = /Users/runner/work/SDL/SDL/dist
>   actions         = {'xcframework'}
>   dry             = False
>   force           = False
>   cmake_generator = Ninja
> Command line invocation:
>     /Applications/Xcode_14.2.app/Contents/Developer/usr/bin/xcodebuild -project /private/tmp/tardir/SDL3-3.0.0/Xcode/SDL/SDL.xcodeproj -target SDL3.dmg -configuration Release
> 
> User defaults from command line:
>     IDEPackageSupportUseBuiltinSCM = YES
> 
> 2024-03-23 03:07:57.320 xcodebuild[966:4846] CFPropertyListCreateFromXMLData(): Old-style plist parser: missing semicolon in dictionary on line 2636. Parsing will be abandoned. Break on _CFPropertyListMissingSemicolon to debug.
> 2024-03-23 03:07:57.322 xcodebuild[966:4846] Error Domain=NSCocoaErrorDomain Code=3840 "JSON text did not start with array or object and option to allow fragments not set. around line 1, column 0." UserInfo={NSDebugDescription=JSON text did not start with array or object and option to allow fragments not set. around line 1, column 0., NSJSONSerializationErrorIndex=0}
> 2024-03-23 03:07:57.326 xcodebuild[966:4846] Writing error result bundle to /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/ResultBundle_2024-23-03_03-07-0057.xcresult
> xcodebuild: error: Unable to read project 'SDL.xcodeproj' from folder '/private/tmp/tardir/SDL3-3.0.0/Xcode/SDL'.
> 	Reason: The project ‘SDL’ is damaged and cannot be opened due to a parse error. Examine the project file for invalid edits or unresolved source control conflicts.
> 
> Path: /tmp/tardir/SDL3-3.0.0/Xcode/SDL/SDL.xcodeproj
> ```
> 
> This is the only diff applied to the Xcode project:
> 
> ```diff
> diff --git a/Xcode/SDL/SDL.xcodeproj/project.pbxproj b/Xcode/SDL/SDL.xcodeproj/project.pbxproj
> --- a/Xcode/SDL/SDL.xcodeproj/project.pbxproj
> +++ b/Xcode/SDL/SDL.xcodeproj/project.pbxproj
> @@ -2633,7 +2633,7 @@
>                         );
>                         runOnlyForDeploymentPostprocessing = 0;
>                         shellPath = /bin/sh;
> -                       shellScript = "mkdir -p build/dmg-tmp\ncp -a build/SDL3.xcframework build/dmg-tmp/\n\ncp pkg-support/resources/License.txt build/dmg-tmp\ncp pkg-support/resources/ReadMe.txt build/dmg-tmp\n\n# remove the .DS_Store files if any (we may want to provide one in the future for fancy .dmgs)\nfind build/dmg-tmp -name .DS_Store -exec rm -f \"{}\" \\;\n\n# for fancy .dmg\nmkdir -p build/dmg-tmp/.logo\ncp pkg-support/resources/SDL_DS_Store build/dmg-tmp/.DS_Store\ncp pkg-support/sdl_logo.pdf build/dmg-tmp/.logo\n\n# create the dmg\nhdiutil create -ov -fs HFS+ -volname SDL3 -srcfolder build/dmg-tmp build/SDL3.dmg\n\n# clean up\nrm -rf build/dmg-tmp\n";
> +                       shellScript = "mkdir -p build/dmg-tmp/share/cmake/SDL3\ncp -a build/SDL3.xcframework build/dmg-tmp/\n\ncp pkg-support/resources/License.txt build/dmg-tmp\ncp pkg-support/resources/ReadMe.txt build/dmg-tmp\ncp pkg-support/resources/CMake/sdl3-config.cmake build/dmg-tmp/share/cmake/SDL3\ncp pkg-support/resources/SDL3/sdl3-config-version.cmake build/dmg-tmp/share/cmake/SDL3\n\n\nmkdir -p \n\n# remove the .DS_Store files if any (we may want to provide one in the future for fancy .dmgs)\nfind build/dmg-tmp -name .DS_Store -exec rm -f "{}" \\;\n\n# for fancy .dmg\nmkdir -p build/dmg-tmp/.logo\ncp pkg-support/resources/SDL_DS_Store build/dmg-tmp/.DS_Store\ncp pkg-support/sdl_logo.pdf build/dmg-tmp/.logo\n\n# create the dmg\nhdiutil create -ov -fs HFS+ -volname SDL3 -srcfolder build/dmg-tmp build/SDL3.dmg\n\n# clean up\nrm -rf build/dmg-tmp\n";
>                 };
>                 F3B38CF0296F63D1005DA6D3 /* ShellScript */ = {
>                         isa = PBXShellScriptBuildPhase;
> ```

I think you need to escape the quotes used in the argument to `rm`

**Comment by madebr** [comment]
I have it in the green again: https://github.com/madebr/SDL/actions/runs/8403796563

The workflow was started by running:
```
gh workflow run release.yml --ref releaser -f commit=$(git rev-parse releaser)
```
- `--ref releaser` is the branch that contains the `releaser.py` script
- `-f commit=.......` is the input for the `releaser.py` script. The script will create release artifacts from that hash.

## remarks about the script/process
- Source archives are created from the local git repo using `git ls-tree`, `git cat-file` and `git log`. This way, you are absolutely sure no local edits or files accidentally appear in a release.
- The source archives have the `.github` directory filter out: I don't think that belongs in a source release
- The source archives will also contain a `VERSION.txt` file and `.git-hash` file.
- The releaser.yml workflow is a `workflow_dispatch` workflow, which means it must be manually started (using e.g. `gh workflow run ...`)
- The binary archives are created from the source archive in separate independent jobs


Creating of the source archives is relatively slow on GitHub actions because it runs 2 interacting processes in parallel: `releaser.py` and `git cat-file`. It looks like the job nature of GitHub ci is not made for interactive processes.
This slowless does not manifest on my dev machine.

**Comment by slouken** [comment]
This is perfect timing, I'm planning to release an SDL 3.0 preview tomorrow. Go ahead and merge it and I'll try out the flow in the morning and we can tweak it if needed.

**Comment by madebr** [comment]
> This is perfect timing, I'm planning to release an SDL 3.0 preview tomorrow. Go ahead and merge it and I'll try out the flow in the morning and we can tweak it if needed.

I didn't document the script in every detail, but I think the sript is self-explanatory. Use `--help` to know what each option does.
These are the highlights:

Create a source release:
```
build-scripts/releaser.py --create source --project SDL3
```
Replace "source" with "win32" for a MSVC release, replace with "mingw" for a mingw release and with "xcframework" for the dmg
Add `--force` to force things and also to ignore a dirty git workspace.
I added `--dry-run` to test the MSVC/framework flow on Linux (it creates dummy files instead).
It places generated files in a "dist" folder.

With some modifications, the script can also be used to create releases for the satellite libraries. That's why the script has a `--project` argument.

**Comment by slouken** [comment]
It looks like the github workflow has the ability to verify the packages, which is great. Is there a GitHub command to run the workflow, build all packages, and verify them?

**Comment by madebr** [comment]
This one:
```sh
gh workflow run release.yml --ref main -f commit=GIT_HASH_YOU_WISH_TO_BUILD_A_RELEASE_FOR
```
- starts the `workflow_dispatch` workflow, present on the `main` branch
- uses the `releaser.py` script of the `main` branch
- builds a release for the hash `GIT_HASH_YOU_WISH_TO_BUILD_A_RELEASE_FOR`

During development, I did `-f commit=$(git rev-parse releaser)` to get the hash of the releaser branch.

**Comment by slouken** [comment]
Where are the build outputs?

**Comment by madebr** [comment]
> Where are the build outputs?

They appear at the bottom of the workflow summary.
https://github.com/libsdl-org/SDL/actions/runs/8406215437

**Comment by slouken** [comment]
This is very cool, thank you!

I noticed the Apple verify build has this warning:
```
CMake Warning (dev) at /Volumes/SDL3/share/cmake/SDL3/sdl3-config-version.cmake:14 (message):
  Cannot not find SDL_version.h.  This script is meant to be placed in
  share/cmake/SDL3, next to SDL3.xcframework
```
Is that an issue?

**Comment by madebr** [comment]
Oh, I didn't see those before.
It means people can't require a minimum SDL version. I'll fix it.

**Pull request merged**